File: app/src/main/java/com/itachi1706/busarrivalsg/MainMenuActivity.java
package com.itachi1706.busarrivalsg;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.NotificationChannel;
import android.app.NotificationChannelGroup;
import android.app.NotificationManager;
import android.app.ProgressDialog;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.google.firebase.crashlytics.FirebaseCrashlytics;
import com.itachi1706.appupdater.AppUpdateInitializer;
import com.itachi1706.appupdater.object.CAAnalytics;
import com.itachi1706.appupdater.utils.AnalyticsHelper;
import com.itachi1706.busarrivalsg.AsyncTasks.GetAllBusStops;
import com.itachi1706.busarrivalsg.AsyncTasks.GetBusServicesFavouritesRecycler;
import com.itachi1706.busarrivalsg.Database.BusStopsDB;
import com.itachi1706.busarrivalsg.RecyclerViews.FavouritesRecyclerAdapter;
import com.itachi1706.busarrivalsg.Services.BusStorage;
import com.itachi1706.busarrivalsg.objects.BusServices;
import com.itachi1706.busarrivalsg.util.LogInitializer;
import com.itachi1706.busarrivalsg.util.StaticVariables;
import com.itachi1706.busarrivalsg.util.SwipeFavouriteCallback;
import com.itachi1706.busarrivalsg.util.SwipeMoveFavouriteCallback;
import com.itachi1706.helperlib.helpers.ConnectivityHelper;
import com.itachi1706.helperlib.helpers.LogHelper;

import java.util.ArrayList;
import java.util.concurrent.TimeUnit;

public class MainMenuActivity extends AppCompatActivity implements SwipeRefreshLayout.OnRefreshListener {

    static { AppCompatDelegate.setCompatVectorFromResourcesEnabled(true); }

    //Android Stuff
    private TextView syncState;
    private FloatingActionButton fab;
    private FavouritesRecyclerAdapter adapter;
    private SwipeRefreshLayout swipeToRefresh;

    private SharedPreferences sp;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(!BuildConfig.DEBUG);
        LogInitializer.initLogger();
        setContentView(R.layout.activity_main_menu_recycler);

        fab = findViewById(R.id.add_fab);
        RecyclerView favouritesList = findViewById(R.id.rvFav);
        syncState = findViewById(R.id.firebase_sync_status);

        // Obtain the FirebaseAnalytics instance.
        FirebaseAnalytics mFirebaseAnalytics = FirebaseAnalytics.getInstance(this);
        AnalyticsHelper helper = new AnalyticsHelper(this, true);
        @SuppressLint("WrongThread") CAAnalytics analytics = helper.getData(BuildConfig.DEBUG);
        setAnalyticsData(analytics != null, mFirebaseAnalytics, analytics); // Update Firebase User Properties
        mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.APP_OPEN, null);

        if (favouritesList != null) favouritesList.setHasFixedSize(true);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setOrientation(RecyclerView.VERTICAL);
        favouritesList.setLayoutManager(linearLayoutManager);
        favouritesList.setItemAnimator(new DefaultItemAnimator());

        swipeToRefresh = findViewById(R.id.refresh_favourites);
        if (swipeToRefresh != null) {
            swipeToRefresh.setOnRefreshListener(this);
            swipeToRefresh.setColorSchemeResources(
                    R.color.refresh_progress_1,
                    R.color.refresh_progress_2,
                    R.color.refresh_progress_3,
                    R.color.refresh_progress_4);
        }

        sp = PreferenceManager.getDefaultSharedPreferences(this.getApplicationContext());
        sp.edit().putBoolean("cepas_dark_theme", true).apply();
        adapter = new FavouritesRecyclerAdapter(new ArrayList<>(), this, StaticVariables.INSTANCE.useServerTime(sp));
        favouritesList.setAdapter(adapter);

        ItemTouchHelper moveAdapter = new ItemTouchHelper(new SwipeMoveFavouriteCallback(this, new SwipeFavouriteCallback.ISwipeCallback() {
            @Override public boolean getFavouriteState(int position) { return true; } // Always favourited
            @Override public boolean moveFavourite(int oldPosition, int newPosition) { return adapter.moveItem(oldPosition, newPosition); }
            @Override public boolean toggleFavourite(int position) { return adapter.removeFavourite(position); }
        }));
        moveAdapter.attachToRecyclerView(favouritesList);

        LogHelper.d("MainMenu", "Checking for app updates");
        new AppUpdateInitializer(this, sp, R.drawable.notification_icon, StaticVariables.BASE_SERVER_URL, true).setOnlyOnWifiCheck(true).checkForUpdate();
        LogHelper.d("MainMenu", "onCreate complete");

        // Create the Firebase Notification Channel
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) createFirebaseNotifChannel();
    }

    @Override
    public void onResume(){
        super.onResume();

        fab.setOnClickListener(v -> startActivity(new Intent(MainMenuActivity.this, BusStopsTabbedActivity.class)));
        fab.setOnLongClickListener(v -> {
            Toast.makeText(MainMenuActivity.this, R.string.fab_hint_main_menu, Toast.LENGTH_SHORT).show();
            return true;
        });

        //Android Stuff now again :D
        checkIfDatabaseUpdated();

        //Update Favourites
        swipeToRefresh.setRefreshing(true);
        updateFavourites();

        /*switch (sp.getString("companionDevice", "none")) {
            case "none":
            default:  break;
        }*/

        syncState.setClickable(true);
        syncState.setOnClickListener(v -> {
            // TODO: Start Activity for result
            startActivity(new Intent(getApplicationContext(), FirebaseLoginActivity.class));
        });

        invalidateOptionsMenu();
    }

    private void setAnalyticsData(boolean enabled, FirebaseAnalytics firebaseAnalytics, CAAnalytics analytics) {
        firebaseAnalytics.setUserProperty("debug_mode", (enabled) ? analytics.isDebug() + "" : null);
        firebaseAnalytics.setUserProperty("device_manufacturer", (enabled) ? analytics.getdManufacturer() : null);
        firebaseAnalytics.setUserProperty("device_codename", (enabled) ? analytics.getdCodename() : null);
        firebaseAnalytics.setUserProperty("device_fingerprint", (enabled) ? analytics.getdFingerprint() : null);
        firebaseAnalytics.setUserProperty("device_cpu_abi", (enabled) ? analytics.getdCPU() : null);
        firebaseAnalytics.setUserProperty("device_tags", (enabled) ? analytics.getdTags() : null);
        firebaseAnalytics.setUserProperty("app_version_code", (enabled) ? Long.toString(analytics.getAppVerCode()) : null);
        firebaseAnalytics.setUserProperty("android_sec_patch", (enabled) ? analytics.getSdkPatch() : null);
        firebaseAnalytics.setUserProperty("AndroidOS", (enabled) ? Integer.toString(analytics.getSdkver()) : null);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main_menu, menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
        menu.findItem(R.id.ntu_tracker).setVisible(sp.getBoolean("showntushuttle", false));

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.action_settings) startActivity(new Intent(this, MainSettings.class));
        else if (id == R.id.view_all_stops) startActivity(new Intent(this, ListAllBusStopsActivity.class));
        else if (id == R.id.action_refresh) {
            swipeToRefresh.setRefreshing(true);
            updateFavourites();
        } else if (id == R.id.ntu_tracker) startActivity(new Intent(this, NTUBusActivity.class));
        else if (id == R.id.scan_cepas) startActivity(new Intent(this, CEPASScanActivity.class));
        else return super.onOptionsItemSelected(item);

        return true;
    }

    private void updateFavourites(){
        final String TAG = "FAVOURITES";
        //Populate favourites from favourites list
        LogHelper.d(TAG, "Favourites Pref: " + sp.getString("stored", "wot"));

        if (BusStorage.hasFavourites(sp)) {
            //Go ahead with loading and getting data
            LogHelper.d(TAG, "Has Favourites. Processing");
            StaticVariables.INSTANCE.setFavouritesList(BusStorage.getStoredBuses(sp));
            adapter.updateAdapter(StaticVariables.INSTANCE.getFavouritesList(), null);
            adapter.notifyDataSetChanged();

            LogHelper.d(TAG, "Finished Processing, retrieving estimated arrival data now");
            new GetBusServicesFavouritesRecycler(this, adapter).executeOnExecutor(StaticVariables.INSTANCE.getFavouritesList().toArray(new BusServices[0]));
            LogHelper.d(TAG, "Finished creating AsyncTasks to retrieve estimated arrival data");
        }

        if (swipeToRefresh.isRefreshing()){
            swipeToRefresh.setRefreshing(false);
        }
    }

    private void checkIfDatabaseUpdated(){
        long busDBLastUpdate = sp.getLong("busDBTimeUpdated", -1);
        boolean busDBUpdate = false;
        if (busDBLastUpdate != -1){
            long day = TimeUnit.MILLISECONDS.toDays(System.currentTimeMillis() - busDBLastUpdate);
            LogHelper.d("INIT", "Bus DB Last Update: " + day);
            if (day > 30)
                busDBUpdate = true;
        }

        // Check upgrade
        final String DBTAG = "DB UPGRADE";
        int dbver = sp.getInt("busDBVerCheck", 0);
        LogHelper.d(DBTAG, "Current DB Version: " + dbver);
        switch (dbver) {
            case 0:
            case 1: LogHelper.i(DBTAG, "Upgrading to V2 API DB"); busDBUpdate = true; sp.edit().putInt("busDBVerCheck", 2).apply(); break;
            case 2: LogHelper.i(DBTAG, "Upgrading to DB with Bus Services"); busDBUpdate = true; sp.edit().putInt("busDBVerCheck", 3).apply(); break;
        }

        //Main Database
        if (!sp.getBoolean("busDBLoaded", false) || busDBUpdate){
            //First Boot, populate database
            if (!ConnectivityHelper.hasInternetConnection(getApplicationContext())) {
                networkUnavailable(getString(R.string.database_name_bus));
            } else {
                LogHelper.d("INIT", "Initializing Bus Stop Database");
                ProgressDialog dialog = new ProgressDialog(this);
                dialog.setTitle(getString(R.string.database_name_bus));
                dialog.setMessage(getString(R.string.dialog_message_retrieve_data_from_server));
                dialog.setCancelable(false);
                dialog.setIndeterminate(true);
                dialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);

                BusStopsDB db = new BusStopsDB(this);
                db.dropAndRebuildDB();
                sp.edit().putBoolean("busDBLoaded", false).apply();
                dialog.show();

                new GetAllBusStops(dialog, db, this, sp).executeOnExecutor(0);
            }
        } else {
            //Legacy Check
            if (sp.getLong("busDBTimeUpdated", -1) == -1){
                sp.edit().putLong("busDBTimeUpdated", System.currentTimeMillis()).apply();
            }
        }
    }

    private void networkUnavailable(String reason){
        new AlertDialog.Builder(this).setTitle(R.string.dialog_title_no_internet)
                .setMessage(getString(R.string.dialog_message_no_internet, reason)).setCancelable(false)
                .setNeutralButton(R.string.dialog_action_neutral_override, null)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> MainMenuActivity.this.finish()).show();
    }

    @Override
    public void onRefresh() {
        updateFavourites();
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private void createFirebaseNotifChannel() {
        NotificationManager notificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        if (notificationManager == null) return;
        NotificationChannel notificationChannel = new NotificationChannel("firebase-msg", "Server Alerts (FB)", NotificationManager.IMPORTANCE_LOW);
        notificationChannel.enableLights(true);
        notificationChannel.setLightColor(Color.RED);
        notificationChannel.enableVibration(true);
        notificationChannel.setVibrationPattern(new long[]{100, 200, 300, 400, 500, 400, 300, 200, 400});
        notificationChannel.setGroup("server-msg");

        NotificationChannelGroup notificationChannelGroup = new NotificationChannelGroup("server-msg", "Server Messages");
        notificationManager.createNotificationChannelGroup(notificationChannelGroup);
        notificationManager.createNotificationChannel(notificationChannel);
    }
}

File: app/src/main/java/com/itachi1706/busarrivalsg/MainMenuActivity.kt
package com.itachi1706.busarrivalsg

import android.content.Intent
import android.content.SharedPreferences
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.annotation.WorkerThread
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.app.NotificationChannelCompat
import androidx.core.app.NotificationChannelGroupCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import androidx.core.content.edit
import androidx.lifecycle.lifecycleScope
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.itachi1706.appupdater.AppUpdateInitializer
import com.itachi1706.appupdater.`object`.CAAnalytics
import com.itachi1706.appupdater.utils.AnalyticsHelper
import com.itachi1706.busarrivalsg.adapters.FavouritesRecyclerAdapter
import com.itachi1706.busarrivalsg.database.BusStopsDb
import com.itachi1706.busarrivalsg.databinding.ActivityMainMenuRecyclerBinding
import com.itachi1706.busarrivalsg.services.BusStorage
import com.itachi1706.busarrivalsg.tasks.GetBusServicesFavouritesTask
import com.itachi1706.busarrivalsg.tasks.UpdateDatabaseTask
import com.itachi1706.busarrivalsg.util.LogInitializer
import com.itachi1706.busarrivalsg.util.StaticVariables
import com.itachi1706.busarrivalsg.util.SwipeFavouriteCallback
import com.itachi1706.busarrivalsg.util.SwipeMoveFavouriteCallback
import com.itachi1706.helperlib.helpers.ConnectivityHelper
import com.itachi1706.helperlib.helpers.EdgeToEdgeHelper
import com.itachi1706.helperlib.helpers.LogHelper
import com.itachi1706.helperlib.helpers.PrefHelper
import kotlinx.coroutines.async
import java.util.concurrent.TimeUnit

class MainMenuActivity : AppCompatActivity(), SwipeRefreshLayout.OnRefreshListener {

    private lateinit var sp: SharedPreferences
    private var busStorage: BusStorage? = null

    private lateinit var binding: ActivityMainMenuRecyclerBinding
    private var adapter: FavouritesRecyclerAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        FirebaseCrashlytics.getInstance().isCrashlyticsCollectionEnabled = !BuildConfig.DEBUG
        LogInitializer.initLogger()

        binding = ActivityMainMenuRecyclerBinding.inflate(layoutInflater)
        EdgeToEdgeHelper.setEdgeToEdgeWithContentView(binding.root, this)
        setSupportActionBar(binding.toolbar)

        // Obtain Firebase Analytics instance
        val analyticsTask = lifecycleScope.async { loadAnalytics() }
        analyticsTask.invokeOnCompletion { _ -> LogHelper.d(TAG, "Firebase Analytics initialization complete") }

        val llm = LinearLayoutManager(this)
        llm.orientation = RecyclerView.VERTICAL

        binding.refreshFavourites.let {
            it.setOnRefreshListener(this)
            it.setColorSchemeResources(
                R.color.refresh_progress_1,
                R.color.refresh_progress_2,
                R.color.refresh_progress_3,
                R.color.refresh_progress_4
            )
        }

        sp = PreferenceManager.getDefaultSharedPreferences(this)
        PrefHelper.handleDefaultThemeSwitch(sp.getString("app_theme", "default") ?: "default")
        sp.edit { putBoolean("cepas_dark_theme", true) }
        adapter = FavouritesRecyclerAdapter(arrayListOf(), this, StaticVariables.useServerTime(sp))
        binding.rvFav.let {
            it.setHasFixedSize(true)
            it.layoutManager = llm
            it.itemAnimator = DefaultItemAnimator()
            it.adapter = adapter
        }

        busStorage = BusStorage(sp)

        val moveAdapter = ItemTouchHelper(SwipeMoveFavouriteCallback(this, object: SwipeFavouriteCallback.ISwipeCallback {
            override fun getFavouriteState(position: Int): Boolean {
                return true // Always favourite here
            }

            override fun moveFavourite(oldPosition: Int, newPosition: Int): Boolean {
                return adapter?.moveItem(oldPosition, newPosition) ?: false
            }

            override fun toggleFavourite(position: Int): Boolean {
                return adapter?.removeFavourite(position) ?: false
            }
        }))
        moveAdapter.attachToRecyclerView(binding.rvFav)

        if (savedInstanceState == null) {
            LogHelper.d(TAG, "Checking for app updates...")
            AppUpdateInitializer(this, sp, R.drawable.notification_icon, StaticVariables.BASE_SERVER_URL, true)
                .setOnlyOnWifiCheck(true).setPathBasedApi(true).checkForUpdate()
            LogHelper.d(TAG, "onCreate complete")
        } else {
            LogHelper.d(TAG, "Skipping app update check as it should be already done")
        }


        // Create Firebase Notification Channel
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            createFirebaseNotifChannel()
        }
    }

    @WorkerThread
    fun loadAnalytics() {
        val mFirebaseAnalytics = FirebaseAnalytics.getInstance(this)
        val helper = AnalyticsHelper(this, true)
        val analytics = helper.getData(BuildConfig.DEBUG)
        setAnalyticsData(analytics != null, mFirebaseAnalytics, analytics)
        mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.APP_OPEN, null)
        LogHelper.d(TAG, "Firebase Analytics data set")
    }

    override fun onResume() {
        super.onResume()

        binding.addFab.let {
            it.setOnClickListener { startActivity(Intent(this, BusStopsTabbedActivity::class.java)) }
            it.setOnLongClickListener { _ ->
                Toast.makeText(this, R.string.fab_hint_main_menu, Toast.LENGTH_SHORT).show()
                return@setOnLongClickListener true
            }
        }

        checkIfDatabaseUpdated()

        binding.refreshFavourites.isRefreshing = true
        updateFavourites()

        // Companion objects
        when (sp.getString("companionDevice", "none")) {
            "pebble" -> LogHelper.i(TAG, "Pebble selected but deprecated. No action taken")
            else -> LogHelper.i(TAG, "No companion device selected")
        }

        binding.firebaseSyncStatus.let {
            it.isClickable = true
            it.setOnClickListener { startActivity(Intent(this, FirebaseLoginActivity::class.java)) }
        }

        invalidateOptionsMenu()
    }

    private fun setAnalyticsData(enabled: Boolean, firebaseAnalytics: FirebaseAnalytics, analytics: CAAnalytics?) {
        firebaseAnalytics.setUserProperty("debug_mode", if (enabled) analytics?.isDebug.toString() else null)
        firebaseAnalytics.setUserProperty("device_manufacturer", if (enabled) analytics?.getdManufacturer() else null)
        firebaseAnalytics.setUserProperty("device_codename", if (enabled) analytics?.getdCodename() else null)
        firebaseAnalytics.setUserProperty("device_fingerprint", if (enabled) analytics?.getdFingerprint() else null)
        firebaseAnalytics.setUserProperty("device_cpu_abi", if (enabled) analytics?.getdCPU() else null)
        firebaseAnalytics.setUserProperty("device_tags", if (enabled) analytics?.getdTags() else null)
        firebaseAnalytics.setUserProperty("app_version_code", if (enabled) analytics?.appVerCode.toString() else null)
        firebaseAnalytics.setUserProperty("android_sec_patch", if (enabled) analytics?.sdkPatch else null)
        firebaseAnalytics.setUserProperty("AndroidOS", if (enabled) analytics?.sdkver.toString() else null)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main_menu, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        super.onPrepareOptionsMenu(menu)
        menu.findItem(R.id.ntu_tracker).isVisible = sp.getBoolean("showntushuttle", false)

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_settings -> {
                startActivity(Intent(this, MainSettings::class.java))
                true
            }
            R.id.view_all_stops -> {
                startActivity(Intent(this, ListAllBusStopsActivity::class.java))
                true
            }
            R.id.action_refresh -> {
                binding.refreshFavourites.isRefreshing = true
                updateFavourites()
                true
            }
            R.id.ntu_tracker -> {
                startActivity(Intent(this, NtuBusActivity::class.java))
                true
            }
            R.id.scan_cepas -> {
                startActivity(Intent(this, CEPASScanActivity::class.java))
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun updateFavourites() {
        val innerTag = "FAVOURITES:"
        // Populate favourites
        LogHelper.d(TAG, "$innerTag Favourited Pref: ${sp.getString("stored", "wot")}")

        if (busStorage?.hasFavourites() ?: false) {
            // Load and get data
            LogHelper.d(TAG, "$innerTag Has favourites, processing...")
            StaticVariables.favouritesList = ArrayList(busStorage?.getStoredBuses() ?: emptyList())
            adapter?.let {
                it.updateAdapter(StaticVariables.favouritesList, null)
                it.notifyItemRangeChanged(0, StaticVariables.favouritesList.size)

                LogHelper.d(TAG, "$innerTag Finished Processing, retrieving estimated arrival data...")
                GetBusServicesFavouritesTask(this, it).executeOnExecutor(*StaticVariables.favouritesList.toTypedArray())
                LogHelper.d(TAG, "$innerTag AsyncTask created to retrieved estimated arrival data")
            }
        }

        binding.refreshFavourites.let {
            if (it.isRefreshing) it.isRefreshing = false
        }
    }

    private fun checkIfDatabaseUpdated() {
        val busDbLastUpdate = sp.getLong(BUS_DB_TIME_UPDATE_CHECK, -1)
        var busDbUpdate = false
        if (busDbLastUpdate != -1L) {
            val day = TimeUnit.MILLISECONDS.toDays(System.currentTimeMillis() - busDbLastUpdate)
            LogHelper.d(TAG, "Init: Days since Bus DB last update: $day")
            if (day > 30) {
                busDbUpdate = true
            }
        }

        // Check for upgrades
        val dbTag = "DB UPGRADE"
        val dbVer = sp.getInt(DB_VERSION_CHECK, 0)
        LogHelper.d(TAG, "$dbTag: Current DB Version: $dbVer")
        when (dbVer) {
            0, 1 -> {
                // Upgrade to 2
                LogHelper.i(TAG, "$dbTag: Upgrading to V2 API DB")
                busDbUpdate = true
                sp.edit { putInt(DB_VERSION_CHECK, 2) }
            }
            2 -> {
                // Upgrade to 3
                LogHelper.i(TAG, "$dbTag: Upgrading to DB with Bus Services")
                busDbUpdate = true
                sp.edit { putInt(DB_VERSION_CHECK, 3) }
            }
            else -> LogHelper.d(TAG, "$dbTag: No DB upgrade required")
        }

        // Main DB
        if (!sp.getBoolean("busDBLoaded", false) || busDbUpdate) {
            // First boot, populate DB
            if (!ConnectivityHelper.hasInternetConnection(applicationContext)) {
                networkUnavailable(getString(R.string.database_name_bus))
            } else {
                LogHelper.d(TAG, "Init: Initializing Bus Stop Database")
                if (busDbUpdate) {
                    BusStopsDb(this).use {
                        it.dropAndRebuildDb() // Rebuild DB
                        LogHelper.i(TAG, "$dbTag: Database upgraded")
                    }
                }

                ContextCompat.startForegroundService(this, Intent(this, UpdateDatabaseTask::class.java))
            }
        } else {
            // Legacy Check
            if (sp.getLong(BUS_DB_TIME_UPDATE_CHECK, -1) == -1L) {
                sp.edit { putLong(BUS_DB_TIME_UPDATE_CHECK, System.currentTimeMillis()) }
            }
        }
    }

    private fun networkUnavailable(reason: String) {
        MaterialAlertDialogBuilder(this).setTitle(R.string.dialog_title_no_internet)
            .setMessage(getString(R.string.dialog_message_no_internet, reason))
            .setCancelable(false)
            .setNeutralButton(R.string.dialog_action_neutral_override, null)
            .setPositiveButton(android.R.string.ok) { _, _ -> finish() }
            .show()
    }

    override fun onRefresh() {
        updateFavourites()
    }

    private fun createFirebaseNotifChannel() {
        val notificationManager = NotificationManagerCompat.from(this)
        val notificationChannel = NotificationChannelCompat.Builder("firebase-msg", NotificationManagerCompat.IMPORTANCE_LOW)
            .setName("Server Alerts (FB)").setLightsEnabled(true)
            .setLightColor(Color.RED)
            .setVibrationEnabled(true)
            .setVibrationPattern(longArrayOf(100, 200, 300, 400, 500, 400, 300, 200, 400))
            .setGroup("server-msg")
            .build()

        val notificationChannelGroup = NotificationChannelGroupCompat.Builder("server-msg")
            .setName("Server Messages")
            .build()

        notificationManager.createNotificationChannelGroup(notificationChannelGroup)
        notificationManager.createNotificationChannel(notificationChannel)
    }

    companion object {
        init {
            AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        }

        private const val DB_VERSION_CHECK = "busDBVerCheck"
        private const val BUS_DB_TIME_UPDATE_CHECK = "busDBTimeUpdated"

        private const val TAG = "MainMenuActivity"
    }
}